### ISSUE TITLE
Codegen compilation error for Float scalars with default values in Input objects

### ISSUE THREAD

**Issue by joaquim-verges** [apollographql/apollo-android#3108]
**Summary**

We recently upgraded from Apollo Android 2.4.1 to 2.5.6, and encountered a compilation error in the autogenerated code.

In short, Input objects in our `json` schema with inputs of type `Float` and default values that omit the decimal part (eg. `"defaultValue": "1"`) causes the Java codegen to generate a default value as an `Integer` instead of `Double` which causes a compilation error:

```
private Input<Double> myFloatField = Input.fromNullable(1);
                                                        ^
equality constraints: Double
lower bounds: Integer
```

**Version**
2.5.6

**Description**

I was able to narrow down the issue to a specific codegen corner case:

- Only happens in Java codegen (issue does not happen with Kotlin models)
- Only happens with `json` schema format (issue does not happen with `sdl` schema format)

When a `INPUT_OBJECT` type has a `Float` input field with a default value of `1` or `0` (omitting the decimal part) like so:

```json
 {
  "kind": "INPUT_OBJECT",
  "name": "ReviewRefInput",
  "description": "Circle ref to review input",
  "fields": null,
  "inputFields": [
    {
      "name": "myFloatField",
      "description": "Testing float input codegen",
      "type": {
        "kind": "SCALAR",
        "name": "Float",
        "ofType": null
      },
      "defaultValue": "1"
   }
  ]
}
```

The corresponding output in the generated java file is:

```java
private Input<Double> myFloatField = Input.fromNullable(1);
```

Which is a compilation error, as `1` is considered an `Integer`, and therefore that assignment fails with:

```
private Input<Double> myFloatField = Input.fromNullable(1);
                                                        ^
equality constraints: Double
lower bounds: Integer
```

The equivalent Kotlin generated code is correct however:

```kotlin
val myFloatField: Input<Double> = Input.optional(1.0)
```

The Kotlin codegen correctly converts `1` to `1.0` since the scalar is type `Float`, but the Java codegen does not.

This also happens for mandatory fields:

java: `private double mandatoryFloat = 1;`
kotlin: `val mandatoryFloat: Double = 1.0`

Interestingly, the java codegen does the correct transformation when using schemas in `sdl` format, so the issue only happens when parsing from `json`.

I was able to reproduce the issue easily with the `apollo-compiler` unit tests by modifying the `colorInput` entry in `src/test/validation/schema.json`, changing the `green` input field to have a default value of `0` instead of `0.0`.

The generated file `ColorInput.java` shows the error:

`private Input<Double> green = Input.fromNullable(0);`

while the equivalent kotlin file is correct:

`val green: Input<Double> = Input.optional(0.0)`

Hope this is clear enough, let me know if you need any more information, and thanks for this amazing library!

**Comment by joaquim-verges** [comment]
A little more information:

We first checked if this was a schema definition issue, but the [spec](http://spec.graphql.org/draft/#sec-Float.Input-Coercion) says it is a valid case, and we confirmed it [here](https://github.com/graphql/graphql-spec/issues/868).

I looked a bit deeper at the code, and believe the issue was introduced in #2917 specifically in `GraphQLParser.ValueContext.toKotlin()` that was added to `IRBuilder.kt` l. 152:

```kotlin
is GraphQLParser.NumberValueContext -> if (
          NUMBER().text.contains('.') ||
          NUMBER().text.contains("e")) {
        NUMBER().text.toDouble()
      } else {
        NUMBER().text.toInt()
      }
```

Which shows exactly what the problem is.

I was able to fix it by passing the `IntrospectionSchema.TypeRef` to that method and check the scalar type to return the correct type. 

I'm going to push a PR for it :)
